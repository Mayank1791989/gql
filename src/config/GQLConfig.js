/* @flow */
/* eslint-disable no-use-before-define */
import findConfig from 'find-config';
import JSON5 from 'json5';
import fs from 'fs';
import minimatch from 'minimatch';
import parseGlob from 'parse-glob';
import path from 'path';
import _uniq from 'lodash/uniq';

type Options = {
  cwd?: string
};

const CONFIG_FILE_NAME = '.gqlconfig';

export type QueryParser = (
  'QueryParser'
  | ['EmbeddedQueryParser', { startTag: string, endTag: string }]
);
export type Globs = string | Array<string>;
export type FileMatchConfig = Globs | { include: Globs, ignore?: Globs };
export type ValidateConfig = {
  extends: string,
  rules?: {
    [ruleName: string]: 'off' | 'warn' | 'error',
  },
};
export type QueryConfig = {
  match: FileMatchConfig,
  parser: QueryParser,
  isRelay?: boolean,
  validate?: ValidateConfig,
};

type GQLConfigFile = {
  schema: {
    files: FileMatchConfig,
    validate?: ValidateConfig,
  },
  query?: { // query optional
    files: Array<QueryConfig>,
  },
};

export type { Options as GQLConfigOptions };

class GQLConfig {
  _path: string;
  _configObj: GQLConfigFile;
  _dir: string;

  constructor(options: ?Options) {
    const result = findConfig.obj(CONFIG_FILE_NAME, options);
    if (!result) {
      // throw error .gqlConfig not found
      throw new Error(`Could not find a ${CONFIG_FILE_NAME} file. Create a ${CONFIG_FILE_NAME} file in project root directory.`);
    }

    this._dir = result.dir;
    this._path = result.path;
    this._configObj = this._readConfig(result.path);
  }

  getDir() {
    return this._dir;
  }

  getPath() {
    return this._path;
  }

  getFileExtensions(): Array<string> {
    const { schema, query } = this._configObj;
    const extensions = [];

    // schema
    extensions.push(...extractExtensions(schema.files));

    // query
    if (query) {
      query.files.forEach(({ match }) => {
        extensions.push(...extractExtensions(match));
      });
    }

    return _uniq(extensions);
  }

  getSchema() {
    return this._configObj.schema;
  }

  getQuery() {
    return this._configObj.query;
  }

  match(filePath: string): ?{ type: 'query' | 'schema', opts: Object } {
    const { schema, query } = this._configObj;
    const filePathRelative = path.relative(this.getDir(), filePath);

    // test schema
    if (matcher(filePathRelative, schema.files)) {
      return { type: 'schema', opts: {} };
    }

    // test query
    if (!query) { return null; }

    const { files } = query;
    const opts = files.find(({ match }) => matcher(filePathRelative, match));
    if (opts) {
      return { type: 'query', opts };
    }

    return null;
  }

  _readConfig(filePath: string) { // eslint-disable-line class-methods-use-this
    const fileData = fs.readFileSync(filePath, 'utf8');
    try {
      return (JSON5.parse(fileData): GQLConfigFile); // using flow-runtime it will be validated also
    } catch (err) {
      throw new Error(`\nError parsing ${CONFIG_FILE_NAME}. \n\n${err.message}`);
    }
  }
}

function extractExtensions(match: FileMatchConfig): Array<string> {
  if (typeof match === 'string') {
    return [getExtFromGlob(match)];
  }

  if (Array.isArray(match)) {
    return match.map(getExtFromGlob);
  }

  // object { match, exclude }
  return extractExtensions(match.include);
}

function getExtFromGlob(glob: string): string {
  return parseGlob(glob).path.ext; // without dot
}

function matchGlob(filePath, globs: Globs): boolean {
  if (typeof globs === 'string') {
    return minimatch(filePath, globs);
  }

  // matches any
  return Boolean(globs.find(glob => minimatch(filePath, glob)));
}

function matcher(filePath, match: FileMatchConfig): boolean {
  if (typeof match === 'string' || Array.isArray(match)) {
    return matchGlob(filePath, match);
  }

  const _match = matchGlob(filePath, match.include);
  if (!_match) { return false; }

  const _matchesIgnore = match.ignore ? matchGlob(filePath, match.ignore) : false;

  return _match && !_matchesIgnore;
}

export default GQLConfig;
