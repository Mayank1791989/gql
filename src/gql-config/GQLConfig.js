/* @flow */
import findConfigFile from './findConfigFile';
import readConfigFile from './readConfigFile';
import resolveConfigFile from './resolveConfigFile';
import extractExtensions from './extractExtensions';
import getConfigForFile, {
  type SchemaFileConfig,
  type QueryFileConfig,
} from './getConfigForFile';

import _memoize from 'lodash/memoize';

import { type GQLConfigFileResolved } from './types';
import {
  type ParseOptions as GraphQLParseOptions,
  type BuildSchemaOptions as GraphQLBuildSchemaOptions,
} from 'graphql';

class GQLConfig {
  _dir: string;
  _path: string;
  _configFileResolved: GQLConfigFileResolved;

  constructor(options: { configDir: string }) {
    const result = findConfigFile(options.configDir);
    this._dir = result.dir;
    this._path = result.path;
    this._configFileResolved = resolveConfigFile(
      readConfigFile(result.path),
      result.dir,
    );
    // console.log(JSON.stringify(this._configFileResolved, null, 2));
  }

  // return directory containing config file
  getDir(): string {
    return this._dir;
  }

  // return path of config file
  getPath(): string {
    return this._path;
  }

  // return all file extensions in config file
  getFileExtensions: () => Array<string> = _memoize(() =>
    extractExtensions(this._configFileResolved),
  );

  // return schema config
  getSchemaConfig(): $PropertyType<GQLConfigFileResolved, 'schema'> {
    return this._configFileResolved.schema;
  }

  // return query config
  getQueryConfig(): $PropertyType<GQLConfigFileResolved, 'query'> {
    return this._configFileResolved.query;
  }

  // returns config for given file path
  // file can be schema file or any query file
  getFileConfig(filePath: string): SchemaFileConfig | QueryFileConfig | null {
    return getConfigForFile(filePath, this._configFileResolved, this.getDir());
  }

  getGraphQLParseOptions(): GraphQLParseOptions {
    const { graphQLOptions } = this.getSchemaConfig();
    const options = {};

    if (
      graphQLOptions &&
      typeof graphQLOptions.allowLegacySDLEmptyFields === 'boolean'
    ) {
      options.allowLegacySDLEmptyFields =
        graphQLOptions.allowLegacySDLEmptyFields;
    }

    if (
      graphQLOptions &&
      typeof graphQLOptions.allowLegacySDLImplementsInterface === 'boolean'
    ) {
      options.allowLegacySDLImplementsInterfaces =
        graphQLOptions.allowLegacySDLImplementsInterface;
    }

    return options;
  }

  getGraphQLBuildSchemaOptions(): GraphQLBuildSchemaOptions {
    const { graphQLOptions } = this.getSchemaConfig();
    const options = {};
    if (
      graphQLOptions &&
      typeof graphQLOptions.commentDescriptions === 'boolean'
    ) {
      options.commentDescriptions = graphQLOptions.commentDescriptions;
    }
    return options;
  }
}

export default GQLConfig;
