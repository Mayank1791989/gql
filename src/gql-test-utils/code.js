/* @flow */
import { dedent } from 'dentist';
import splitLines from 'gql-shared/splitLines';

export default function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
