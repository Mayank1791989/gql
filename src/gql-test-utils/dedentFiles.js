/* @flow */
import { dedent } from 'dentist';

type FilesMap = { [path: string]: string };

export function dedentFiles(files?: FilesMap): FilesMap {
  if (!files) {
    return {};
  }
  return Object.keys(files).reduce((acc, path) => {
    acc[path] = dedent(files[path]);
    return acc;
  }, {});
}
