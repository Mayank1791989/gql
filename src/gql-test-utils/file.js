/* @flow */
import os from 'os';
import objectHash from 'object-hash';
import { writeFileSync } from 'gql-shared/fs-utils';
import { execSync } from 'child_process';
import path from 'path';

export const TEST_TEMP_DIR = path.join(os.tmpdir(), 'gql-test');

export function createTempFiles(files: { [file: string]: string }): string {
  // deterministic dirName [as this path used in test cases output (see generated snapshot)]
  const dir = path.join(TEST_TEMP_DIR, objectHash(files));
  Object.keys(files).forEach(file => {
    const content = files[file];
    writeFileSync(path.join(dir, file), content);
  });
  return dir;
}

export function removeTestTempDir() {
  execSync(`rm -rf ${TEST_TEMP_DIR}`);
}
