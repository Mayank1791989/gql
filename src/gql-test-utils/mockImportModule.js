/* @flow */
export default function mockImportModule(
  moduleName: string,
  mockFactory: () => $FixMe,
) {
  jest.doMock('import-from', () => {
    return function importFrom(dir, moduleID) {
      if (moduleID === moduleName) {
        return mockFactory();
      }
      const importFromActual = jest.requireActual('import-from');
      return importFromActual(dir, moduleID);
    };
  });
}
