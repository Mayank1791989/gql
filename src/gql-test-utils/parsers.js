/* @flow */
import SchemaParser from 'gql-schema-parser-default';
import EmbeddedQueryParser from 'gql-query-parser-embedded-queries';
import QueryParser from 'gql-query-parser-default';
import QueryPresetRelay from 'gql-query-preset-relay';

export function relayQLParser() {
  return new EmbeddedQueryParser({
    start: 'Relay\\.QL`',
    end: '`',
    allowFragmentWithoutName: true,
    allowFragmentInterpolation: true,
  });
}

export function apolloParser() {
  return new EmbeddedQueryParser({
    start: 'gql`',
    end: '`',
    allowDocumentInterpolation: true,
  });
}

export const relayExtendSchema = QueryPresetRelay().extendSchema;

export function queryParser() {
  return new QueryParser();
}

export function customParser() {
  return new EmbeddedQueryParser({
    start: '"""',
    end: '"""',
  });
}

export function embeddedQueryParser({
  start,
  end,
}: {
  start: string,
  end: string,
}) {
  return new EmbeddedQueryParser({
    start,
    end,
  });
}

export function schemaParser() {
  return new SchemaParser();
}
