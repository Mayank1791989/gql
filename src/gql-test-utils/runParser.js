/* @flow */
import { type IParser } from 'gql-shared/types';
import MultilineCharacterStream from 'gql-shared/MultilineCharacterStream';

export default function runParser({
  sourceText,
  parser,
}: {
  sourceText: string,
  parser: IParser,
}) {
  const state = parser.startState();
  const stream = new MultilineCharacterStream(sourceText);
  const result = [];
  while (!stream.eof()) {
    const style = parser.token(stream, state);
    result.push({
      start: stream.getStartOfToken(),
      end: stream.getCurrentPosition(),
      text: stream.current(),
      style,
    });
  }

  return result;
}
