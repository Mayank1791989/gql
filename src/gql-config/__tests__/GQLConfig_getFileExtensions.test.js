/* @flow */
import GQLConfig from '../GQLConfig';
import { createTempFiles } from 'gql-test-utils/file';

test('getFileExtensions: detect different file extensions', () => {
  // eslint-disable-next-line no-unused-vars
  const gqlConfig = new GQLConfig({
    configDir: createTempFiles({
      '.gqlconfig': `
          {
            schema: {
              files: ['js/schema/**/*.gql', 'js/schema/**/*.graphql']
            },
            query: {
              files: [
                { match: 'js/files/*.js' },
                {
                  match: [
                    'js/files/*.jsx',
                    'js/file/*.ts',
                    'files/*.go'
                  ]
                },
                {
                  match: {
                    include: 'files/*.py',
                  }
                },
                {
                  match: {
                    include: ['files/*.java', 'files/*.rb'],
                  }
                },
              ]
            }
          }
        `,
    }),
  });

  expect(gqlConfig.getFileExtensions()).toMatchInlineSnapshot(`
Array [
  "gql",
  "graphql",
  "js",
  "jsx",
  "ts",
  "go",
  "py",
  "java",
  "rb",
]
`);
});
