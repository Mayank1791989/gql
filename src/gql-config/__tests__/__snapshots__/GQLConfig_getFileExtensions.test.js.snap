// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFileExtensions: detect different file extensions 1`] = `
Array [
  "gql",
  "graphql",
  "js",
  "jsx",
  "ts",
  "go",
  "py",
  "java",
  "rb",
]
`;
